src/routes/albumRoutes.js

```diff
@@ -10,19 +10,27 @@ const {
   updateAlbum,
   getAllAlbums,
   getSingleAlbum,
+  updatePriceOfAlbums,
   getFilteredAlbums,
+  getAlbumWithAllUsersWhoPurchasedIt,
 } = require('../controllers/albumController');
 
 router
   .route('/')
   .post(authenticateUser, authorizePermissions('admin'), createAlbum) //only admin can create product
-  .get(getAllAlbums); //everyone can access all products- no middleware
+  .get(getAllAlbums) //everyone can access all products- no middleware
+  .patch(authenticateUser, authorizePermissions('admin'), updatePriceOfAlbums);
 
-  router.get('/filter', getFilteredAlbums);
+router.get('/filter', getFilteredAlbums);
 
 router
   .route('/:id')
   .get(getSingleAlbum) //everyone can access all products- no middleware
-  .patch(authenticateUser, authorizePermissions('admin'), updateAlbum); //only admin can update product
+  .get(
```

Comment:
Same thing here about using the same method (`get`) twice on the same route; Express won't know which one to run; so it'll just run the one that was the most recent to be defined. Either put this on a separate route (maybe `/:id/withUsers`) or query parameter on the existing `getSingleAlbum` endpoint

Reply:
@akosasante just did it setting up a separate route  , thank you, I am just curious about how to set up a query parameter on the existing getSingleAlbum endpoint - should we do smth like:

1.  in **albumController.js**

```
const getSingleAlbum = async (req, res) => {
  const { id: albumId } = req.params;
  const listOfUsersWhoPurchasedThisAlbum= req.query.listOfUsersWhoPurchasedThisAlbum=== 'true'; // Check if query parameter is 'true'

  const album = await Album.findOne({ _id: albumId });
  if (!album) {
    throw new CustomError.NotFoundError(`No album with id ${albumId}`);
  }

  let albumResponse = { album };

  if (listOfUsersWhoPurchasedThisAlbum) {
    // Include additional details based on the query parameter
    albumResponse = {
      ...albumResponse,
      additionalDetails: {
        // added additional details here
 let usersThatPurchasedThisAlbum = await Album.findOne({
    _id: req.params.id,
  }).populate({
    path: 'purchasedByUsers', 
    populate: { path: 'user' }, 
  });
        }
    };
  }

  res.status(StatusCodes.OK).json(albumResponse);
};
```

Will it be smth like this? And how would we get from the frontend if user introduced this query parameter? 
With smth like this? -->
 // Construct the URL with query parameters
```
const url = `/albums/${albumId}?listOfUsersWhoPurchasedThisAlbum=${listOfUsersWhoPurchasedThisAlbum}`;

// Make the GET request
fetch(url)
  .then(response => response.json())
  .then(data => {
    // Handle the response data
    console.log(data);
  })
  .catch(error => {
    // Handle errors
    console.error(error);
  });
```

Since only admin can see this info- we need to implement authorizePermissions('admin') middleware anywhere- how would we do it using queries. And what should admin do in browser to get the list of users who bought this album -press some kind of button?

Reply:
I think what you have would work;  but when you do `if(listOfUsersWhoPurchasedThisAlbum)` you could also check the requesting user's role to make sure they are an admin at that point. It would return a bit of redundant data in the `additionalDetails` since it would return a list of albums, then in `additionalDetails` a list of albums _again_ but this time with their users. You could do some `.map` to transform that data into a nicer format.

Here is my thought of how to do it in the backend to avoid that and avoid having to execute two separate queries:

```js
const getSingleAlbum = async (req, res) => {
  const {id: albumId} = req.params;
  // Check if query parameter is the string "true". If not or if the query parameter `?includeUsers` was not passed in at all
  // then just set the variable to false.
  const includePurchasingUsers = req.query.includeUsers === "true" ? true : false;

  // Mongoose will only actually make the query once we call `await`, so we can prep our query first, then actually execute it later
  // https://mongoosejs.com/docs/promises.html
  let albumQuery = Album.findOne({_id: albumId});
  // Only allow admins to view all the users who purchased an album:
  if (includePurchasingUsers && req.user.role === 'admin') {
    albumQuery = albumQuery.populate({path: 'purchasedByUsers', populate: {path: 'user'}});
  }
  // Run the query now
  const album = await albumQuery;

  if (!album) {
    throw new CustomError.NotFoundError(`No album with id ${albumId}`);
  }

  res.status(StatusCodes.OK).json({album});
};
```

For the frontend yes that looks correct to me! 👍 

```
const shouldIncludeUsersWhoPurchasedThisAlbum = true;
const url = `/albums/${albumId}?includeUsers=${shouldIncludeUsersWhoPurchasedThisAlbum}`;

// Make the GET request (I'm showing my example with Axios just cause that's what we're using in the frontend repo; but your fetch example works too)
fetch(url)
  .then(response => response.json())
  .then(data => {
    // Handle the response data
    console.log(data);
  })
  .catch(error => {
    // Handle errors
    console.error(error);
  });
```

I would imagine that either we would have an admin page that only admin can log into that shows this info and makes a request to the backend for this info.

Reply:
@akosasante and using axios will it look like
 ```
const shouldIncludeUsersWhoPurchasedThisAlbum = true;
const url = `/albums/${albumId}?includeUsers=${shouldIncludeUsersWhoPurchasedThisAlbum}`;

// Make the GET request using Axios
axios.get(url)
  .then(response => {
    // Handle the response data
    const data = response.data;
    console.log(data);
  })
  .catch(error => {
    // Handle errors
    console.error(error);
  });
```